exporter/exporterhelper/persistent_queue.go

```diff
@@ -0,0 +1,470 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporterhelper
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/jaegertracing/jaeger/pkg/queue"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/extension/storage"
+)
+
+// persistentQueue holds the queue backed by file storage
+type persistentQueue struct {
+	logger     *zap.Logger
+	stopWG     sync.WaitGroup
+	produceMu  sync.Mutex
+	exitChan   chan struct{}
+	capacity   int
+	numWorkers int
+
+	storage persistentStorage
+}
+
+// persistentStorage provides an interface for request storage operations
+type persistentStorage interface {
+	// put appends the request to the storage
+	put(req request) error
+	// get returns the next available request; not the channel is unbuffered
+	get() <-chan request
+	// size returns the current size of the storage with items waiting for processing
+	size() int
+	// stop gracefully stops the storage
+	stop()
+}
+
+// persistentContiguousStorage provides a persistent queue implementation backed by file storage extension
+//
+// Write index describes the position at which next item is going to be stored.
+// Read index describes which item needs to be read next.
+// When Write index = Read index, no elements are in the queue.
+//
+// The items currently processed by consumers are not deleted until the processing is finished.
+// Their list is stored under a separate key.
+//
+//
+//   ┌───────file extension-backed queue───────┐
+//   │                                         │
+//   │     ┌───┐     ┌───┐ ┌───┐ ┌───┐ ┌───┐   │
+//   │ n+1 │ n │ ... │ 4 │ │ 3 │ │ 2 │ │ 1 │   │
+//   │     └───┘     └───┘ └─x─┘ └─|─┘ └─x─┘   │
+//   │                       x     |     x     │
+//   └───────────────────────x─────|─────x─────┘
+//      ▲              ▲     x     |     x
+//      │              │     x     |     xxxx deleted
+//      │              │     x     |
+//    write          read    x     └── currently processed item
+//    index          index   x
+//                           xxxx deleted
+//
+type persistentContiguousStorage struct {
+	logger                  *zap.Logger
+	client                  storage.Client
+	mu                      sync.Mutex
```

Comment:
Can you document what this mutex is guarding? Ideally it would be place just above the guarded object in the struct